pallets/msa/src/lib.rs

```diff
@@ -556,6 +571,51 @@ pub mod pallet {
 
 			Ok(())
 		}
+
+		/// Retire a MSA
+		///
+		/// ### Events
+		/// - Deposits [`MsaRetired`](Event::MsaRetired) when MSA is retired
+		///
+		/// ### Errors
+		///
+		/// - Returns [`NoKeyExists`](Error::NoKeyExists) if `delegator` does not have an MSA key.
+		/// - Returns [`MoreThanOneKeyExists`](Error::MoreThanOneKeyExists) if the MSA has more than one account key.
+		/// - Returns [`RegisteredProviderCannotBeRetired`](Error::RegisteredProviderCannotBeRetired) if the MSA id is a registered provider
+
+		#[pallet::weight((T::WeightInfo::retire_msa(), DispatchClass::Normal, Pays::No))]
+		pub fn retire_msa(origin: OriginFor<T>) -> DispatchResult {
+			// Check and get the account id from the origin
+			let who = ensure_signed(origin)?;
+
+			// Get the MSA id of the origin
+			let msa_id = Self::get_owner_of(&who).ok_or(Error::<T>::NoKeyExists)?;
```

Comment:
I think there is a function for this `ensure_msa_owner`.

Reply:
We are doing an additional db read here.
I think its already taken care of here : `ensure_msa_owner`.

Reply:
I think you accidentally missed this comment.